packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -688,154 +861,585 @@ class _SortData {
 /// See also:
 ///
 ///  * [FocusNode], for a description of the focus system.
-///  * [DefaultFocusTraversal], a widget that imposes a traversal policy on the
-///    [Focus] nodes below it in the widget hierarchy.
-///  * [WidgetOrderFocusTraversalPolicy], a policy that relies on the widget
+///  * [FocusTraversalGroup], a widget that groups together and imposes a
+///    traversal policy on the [Focus] nodes below it in the widget hierarchy.
+///  * [WidgetOrderTraversalPolicy], a policy that relies on the widget
 ///    creation order to describe the order of traversal.
 ///  * [DirectionalFocusTraversalPolicyMixin] a mixin class that implements
 ///    focus traversal in a direction.
+///  * [OrderedTraversalPolicy], a policy that describes the order
+///    explicitly using [FocusTraversalOrder] widgets.
 class ReadingOrderTraversalPolicy extends FocusTraversalPolicy with DirectionalFocusTraversalPolicyMixin {
-  @override
-  FocusNode findFirstFocus(FocusNode currentNode) {
-    assert(currentNode != null);
-    final FocusScopeNode scope = currentNode.nearestScope;
-    FocusNode candidate = scope.focusedChild;
-    if (candidate == null && scope.traversalChildren.isNotEmpty) {
-      candidate = _sortByGeometry(scope).first;
-    }
-
-    // If we still didn't find any candidate, use the current node as a
-    // fallback.
-    candidate ??= currentNode;
-    candidate ??= FocusManager.instance.rootScope;
-    return candidate;
-  }
-
   // Sorts the list of nodes based on their geometry into the desired reading
   // order based on the directionality of the context for each node.
-  Iterable<FocusNode> _sortByGeometry(FocusScopeNode scope) {
-    final Iterable<FocusNode> nodes = scope.traversalDescendants;
-    if (nodes.length <= 1) {
-      return nodes;
+  @override
+  Iterable<FocusNode> sortDescendants(Iterable<FocusNode> descendants) {
+    assert(descendants != null);
+    if (descendants.length <= 1) {
+      return descendants;
     }
 
-    Iterable<_SortData> inBand(_SortData current, Iterable<_SortData> candidates) {
-      final Rect wide = Rect.fromLTRB(double.negativeInfinity, current.rect.top, double.infinity, current.rect.bottom);
-      return candidates.where((_SortData item) {
-        return !item.rect.intersect(wide).isEmpty;
-      });
+    // Collects the given candidates into groups by directionality. The candidates
+    // have already been sorted as if they all had the directionality of the
+    // nearest Directionality ancestor.
+    List<_ReadingOrderDirectionalGroupData> collectDirectionalityGroups(Iterable<_ReadingOrderSortData> candidates) {
```

Comment:
Can this just be a private method in the class instead of this inline definition? Would make this a little more readable, I think.

Reply:
Same for the others